diff --git a/PATCHED b/PATCHED
new file mode 100644
index 0000000..e69de29
diff --git a/src/model.cc b/src/model.cc
index 0ae3ba4..967fd0c 100644
--- a/src/model.cc
+++ b/src/model.cc
@@ -323,7 +323,7 @@ void Model::ReadDataFiles()
 
         KALDI_LOG << "Loading CARPA model from " << carpa_rxfilename_;
         std_lm_fst_ = fst::ReadFstKaldi(std_fst_rxfilename_);
-        fst::Project(std_lm_fst_, fst::ProjectType::OUTPUT);
+        fst::Project(std_lm_fst_, fst::ProjectType::PROJECT_OUTPUT);
         if (std_lm_fst_->Properties(fst::kILabelSorted, true) == 0) {
             fst::ILabelCompare<fst::StdArc> ilabel_comp;
             fst::ArcSort(std_lm_fst_, ilabel_comp);
